src/compile_target.zig

```diff
@@ -0,0 +1,394 @@
+        }
+    };
+
+    const BaselineFormatter = struct {
+        baseline: bool = false,
+        pub fn format(self: @This(), comptime _: []const u8, _: anytype, writer: anytype) !void {
+            if (self.baseline) {
+                try writer.writeAll("-baseline");
+            }
+        }
+    };
+
+    pub fn eql(this: *const CompileTarget, other: *const CompileTarget) bool {
+        return this.os == other.os and this.arch == other.arch and this.baseline == other.baseline and this.version.eql(other.version) and this.libc == other.libc;
+    }
+
+    pub fn isDefault(this: *const CompileTarget) bool {
+        return this.eql(&.{});
+    }
+
+    pub fn toNPMRegistryURL(this: *const CompileTarget, buf: []u8) ![]const u8 {
+        if (bun.getenvZ("BUN_COMPILE_TARGET_TARBALL_URL")) |url| {
+            if (strings.hasPrefixComptime(url, "http://") or strings.hasPrefixComptime(url, "https://"))
+                return url;
+        }
+
+        return this.toNPMRegistryURLWithURL(buf, "https://registry.npmjs.org") catch unreachable;
+    }
+
+    pub fn toNPMRegistryURLWithURL(this: *const CompileTarget, buf: []u8, registry_url: []const u8) ![]const u8 {
+        return switch (this.os) {
+            inline else => |os| switch (this.arch) {
+                inline else => |arch| switch (this.baseline) {
+                    // https://registry.npmjs.org/@oven/bun-linux-x64/-/bun-linux-x64-0.1.6.tgz
+                    inline else => |is_baseline| try std.fmt.bufPrint(buf, comptime "{s}/@oven/bun-" ++
+                        os.npmName() ++ "-" ++ arch.npmName() ++
+                        (if (is_baseline) "-baseline" else "") ++
+                        "/-/bun-" ++
+                        os.npmName() ++ "-" ++ arch.npmName() ++
+                        (if (is_baseline) "-baseline" else "") ++
+                        "-" ++
+                        "{d}.{d}.{d}.tgz", .{
+                        registry_url,
+                        this.version.major,
+                        this.version.minor,
+                        this.version.patch,
+                    }),
+                },
+            },
+        };
+    }
+
+    pub fn format(this: @This(), comptime _: []const u8, _: anytype, writer: anytype) !void {
+        try std.fmt.format(
+            writer,
+            // bun-darwin-x64-baseline-v1.0.0
+            // This doesn't match up 100% with npm, but that's okay.
+            "bun-{s}-{s}{}{}-v{d}.{d}.{d}",
+            .{
+                this.os.npmName(),
+                this.arch.npmName(),
+                this.libc,
+                BaselineFormatter{ .baseline = this.baseline },
+                this.version.major,
+                this.version.minor,
+                this.version.patch,
+            },
+        );
+    }
+
+    pub fn displayName(this: *const CompileTarget, version_buf: []u8) [:0]const u8 {
+        return std.fmt.bufPrintZ(version_buf, "{}", .{this.*}) catch unreachable;
```

Comment:
It might be caught earlier, but this unreachable might be reachable if we're given huge major/minor/patch values